### ISSUE TITLE
Indentation issue with regard unbalanced brackets and parentheses.

### ISSUE THREAD

**Issue by zhouyan** [lervag/vimtex#592]
Setting `vimtex_indent_enabled` to 1 will be enough to reproduce the following issue.

The `vimtex` indent file indent things between  `[`, `(` and their right counter-parts. However, unlike `{`, they are possibly unbalanced. For example, say an closed-open interval like `[0, 1)` appears often enough. And in some disciplines users will also need to use notions like `] a, b [`. Whenever the delimiters are unbalanced, there appears to be issues with the indentation. For example, consider the following two short paragraph

```
first $[0, 1)$

next
```

The second paragraph and anything that follows will be indented one level. This is quite annoying. Ideally it is best to have the indent expression to analysis these situations. But they do not have to used in math mode. They can be used anywhere. And thus it maybe best not to indent things between them at all.

I understand the where they can be useful. For example, `[` and `]` are commonly use to mark optional arguments. But most of the time, things between them are not long enough to be broken into lines for indentation to be of any use at all. The only one exception is perhaps `\documentclass`. But that hardly is general.

**Comment by lervag** [comment]
Surprisingly, this is the first time someone has raised this issue. I don't really see any good way to solve this, except perhaps by adding some option(s).

The delimiter indentation is actually quite complex. It relies on a delimiter parser that searches for matching delimiters, accounting for e.g. modifiers and similar. It also handles math environments differently from normal text. I don't see how to update this to detect for instance `[ ... )` pairs or `] ... [` pairs used in inline math, because these pairs will typically be "ignored". That is, inline math will typically not be detected, and so the paragraph is detected as "normal" text, which only indents on `[...]` and `{...}` pairs.

One possibility is to add an option that disables indentation on delimiters altogether. One could also disable indentation for delimiters in normal text (`[...]` and `{...}`), but I think this is all suboptimal.

I'm trying to think of a good solution, but it is not clear to me yet.

**Comment by zhouyan** [comment]
Exactly. I don't see anyway it can be done either. In most languages, delimiters have to be balanced except in strings. It makes the situation much easier to parse. However, in tex almost everything is a string. An opening parenthesis in one place is a delimiter while in another place can is just another character. The two places can even be exactly the same and they mean different things only because of the context. So, it is impossible to parse them correctly without some AI the way I see it.

In short, I think it is better to provide an option instead of trying to solve it.

**Comment by zhouyan** [comment]
A another related issue is that suggest to add an option to disable indentation for text with a verbatim environment. Even better would be to have an option that specifies what environments are to be excluded from indentation or any formatting for that matter. It is quite annoying if one indented the whole file and then realized the verbatim text, which are formatted according to the convention of another language, is affected.

**Comment by lervag** [comment]
Verbatim environments are already supported and not indented. This should work. If it doesn't, then please open a new issue and explain the problem.

**Comment by zhouyan** [comment]
Thanks. In that case it is more likely I messed up somewhere in my vimrc.

**Comment by lervag** [comment]
Ok, I've tried to implement a solution. Can you test? Please checkout the branch [improvement/issue-592](/lervag/vimtex/tree/improvement/issue-592) and test.

What I did was to extract the math parts and the normal text parts of each line and to parse these differently. There are still some assumptions, but the updated version passes all tests. However, it is slower than the previous version. I am not sure if that is a big problem, though.

**Comment by zhouyan** [comment]
I tried some basic scenarios and the patch seems to work well.

One possible alternative is to do what LaTeX-Box does. It parses `\(`, `\left(`  but not `(`. The same for `\[`, `\left[` .

If you think it is necessary to parse `[`  in text, I think it is good enough. But I still think it is better to have an option not to parse them at all. It is not very often that one will have things delimited by them long enough to need indentation. The situation where it is useful to have them indented is in math formula, for example,
```
\Bigl[
  % A long formula
`Bigr]
```
On the other hand, within text, they can also be unbalanced. For example,
```
In R, there is the operator \verb|[<-|...
that ...
```
The second line will be indented and all the rest. This is a real example of how such unbalanced brackets can appear even outside math mode.

Again, it is difficult to parse in this case. For example, if one use `shortvrb`, then this `[` can appear between any pair of delimiters.

In summary, the issue is not only with math mode, but the fact that they don't have to be balanced at all in TeX. I think it is why LaTeX-Box only parse `\(` etc. Because they as special construct has to be balanced. So does `\left` and `\right`, though what follows them don't have to be (they an be followed by dot).

The things that in text that has to be balanced are few.
```
{ }
\left \right
```
That's all. Even `\Bigl` etc., don't have to be. For example, say I have a very large formula, starting with `\Biggl`, I only give the first half, followed by some explanation in text mode, and followed by the second have which ends with `\Biggr`. Clearly we shall not indent the text part in between. But it will be indented like the following,
```
\begin{equation}
  [ % ... first half of the large formula
  \end{equation}
  some  text
  \begin{equation}
    % .. continue
  ]
\end{equation}
```
Maybe some people will prefer it this way. But I don't think it is the desired effect in all situations.

**Comment by lervag** [comment]
I actually do things similar to LaTeX-Box, but better. I indent on math delimiters inside math environments (both inline and displayed). However, in normal text, I only indent on `[]` and `{}` pairs.

I think the current implementation handles most cases. Your example for normal mode text is not handled., but it should be relatively easy to add this. I can simply remove all `\verbx...x`s from the normal text before parsing for delimiters.

I think your last example is much more difficult, and it is unclear how to fix it. One could add an editor-specific fix, i.e. some tag in a comment or similar to add/remove an indent. But I don't think this scenario is very common, and so personally I would be happy if I could ignore this...?

**Comment by lervag** [comment]
Regarding `shortvrb`, I really have to say it makes the parsing very much more difficult. I hope you can accept that I _won't_ handle it. If you use `shortvrb` to define `\verb|...|` stuff that introduces unmatched delimiters, than vimtex simply won't handle that.

I've pushed an update that handles the `\verb` commands, though. I feel this should suffice.

**Comment by zhouyan** [comment]
Please don't be bothered too much by my last few examples. I started to encounter the issues because I was writing a document with a lot half-open intervals. Those examples are just to make a point that it is probably a good idea to have the option of not indenting them at all,  because there are too many "reasonable possible" situations in TeX for them to always be dealt with correctly. Both examples were certainly uncommon and don't worth the effort to handle, even though they were real. Even handling \verb is beyond what I hoped for.

Anyway, many thanks for the improvement. I guess you decided not to have an option to have them disabled completely. I think that's fine.

**Comment by lervag** [comment]
Thanks! I think I want to prevent a new option at this moment. If anyone insist that there should be an option, then I'll reconsider, but in the meantime, I think the recent updates should make indentation work in most cases, even some edge cases that are difficult to handle.

I want to clean up the code and also investigate the performance a little bit (I've noticed that the update makes indentation noticably slower) before I merge into master. In the meantime. I'll leave this issue open until I've merged. I don't think it will be too long, and I'll make sure to notify you by closing the issue when merged.
